joblib/memory.py

```diff
@@ -784,9 +793,17 @@ def _persist_input(self, output_dir, duration, args, kwargs,
         metadata = {"duration": duration, "input_args": input_repr}
         try:
             mkdirp(output_dir)
-            with open(os.path.join(output_dir, 'metadata.json'), 'w') as f:
+            # To avoid race conditions with different processes
+            # persisting the same input we write to a temporary
+            # filename before renaming it
+            thread_id = id(threading.current_thread())
+            filename = os.path.join(output_dir, 'metadata.json')
+            temporary_filename = '{}.thread-{}-pid-{}'.format(
+                filename, thread_id, os.getpid())
+            with open(temporary_filename, 'w') as f:
                 json.dump(metadata, f)
-        except:
+            os.rename(temporary_filename, filename)
```

Comment:
You should write a utility helper to factorize the common code for the write + rename pattern.